File: backend/src/services/claude.service.ts
import Anthropic from '@anthropic-ai/sdk';
import { config } from '../config/env';
import { StoryData } from '../types';

const client = new Anthropic({
  apiKey: config.anthropic.apiKey,
});

interface GenerateStoryInput {
  subject: string;
  topic: string;
  question: string;
  type: string;
  options?: any;
  answer: string;
  explanation?: string;
  difficulty: string;
  theme?: string;
}

export class ClaudeService {
  private static buildPrompt(input: GenerateStoryInput): string {
    const optionsText = input.options
      ? `\nOptions: ${JSON.stringify(input.options, null, 2)}`
      : '';

    return `당신은 교육 콘텐츠 전문가이며 인터랙티브 스토리텔링 전문가입니다.

다음 교육 문제를 매력적인 상황극 시나리오로 변환해주세요:

문제 정보:
과목: ${input.subject}
주제: ${input.topic}
난이도: ${input.difficulty}
문제 유형: ${input.type}
질문: ${input.question}${optionsText}
정답: ${input.answer}
${input.explanation ? `설명: ${input.explanation}` : ''}

요구사항:
1. 학생들이 공감할 수 있는 현실적이고 관련성 있는 시나리오를 만드세요
2. 학생들이 연결될 수 있는 캐릭터(이름, 역할, 성격)를 디자인하세요
3. 자연스럽고 연령에 적합한 매력적인 대화를 작성하세요
4. 각 선택지가 문제의 옵션과 명확하게 매핑되도록 하세요
5. 정답과 오답 모두에 대해 건설적인 피드백을 제공하세요
6. 긍정적이고 격려하는 톤을 유지하세요
7. 한국어로 작성하세요
8. 스토리는 2-4개의 장면으로 구성하세요
9. 교육적 가치를 유지하면서 재미있게 만드세요

출력 형식: JSON으로 다음 구조를 따라주세요:
{
  "title": "스토리 제목 (흥미롭고 교육적)",
  "context": "배경 설정 (2-3 문장, 학생을 스토리에 몰입시키기)",
  "character": {
    "name": "캐릭터 이름",
    "role": "캐릭터의 역할",
    "personality": "캐릭터의 성격 (간단히)"
  },
  "scenes": [
    {
      "id": "scene-1",
      "dialogue": "캐릭터의 대사 (자연스럽고 친근하게)",
      "narration": "상황 설명 (학생이 이해하기 쉽게)",
      "choices": [
        {
          "id": "choice-1",
          "text": "선택지 텍스트",
          "isCorrect": true 또는 false,
          "feedback": "선택에 대한 피드백 (긍정적이고 교육적)",
          "nextScene": "scene-2" 또는 null (스토리 종료)
        }
      ]
    }
  ],
  "originalProblem": {
    "question": "${input.question}",
    "answer": "${input.answer}",
    "explanation": "${input.explanation || ''}"
  }
}

중요: 반드시 유효한 JSON만 반환하세요. 추가 설명이나 마크다운 없이 JSON만 출력하세요.

학습을 즐겁게 만드는 몰입형 교육 경험을 만들어주세요!`;
  }

  static async generateStory(input: GenerateStoryInput): Promise<StoryData> {
    const startTime = Date.now();

    try {
      const prompt = this.buildPrompt(input);

      const message = await client.messages.create({
        model: 'claude-sonnet-4-20250514',
        max_tokens: 4096,
        temperature: 0.7,
        messages: [
          {
            role: 'user',
            content: prompt,
          },
        ],
      });

      const content = message.content[0];
      if (content.type !== 'text') {
        throw new Error('Unexpected response type from Claude API');
      }

      const responseText = content.text.trim();

      // Extract JSON from response (handle potential markdown code blocks)
      let jsonText = responseText;
      const jsonMatch = responseText.match(/```json\s*([\s\S]*?)\s*```/);
      if (jsonMatch) {
        jsonText = jsonMatch[1];
      } else if (responseText.startsWith('```') && responseText.endsWith('```')) {
        jsonText = responseText.slice(3, -3).trim();
      }

      const storyData = JSON.parse(jsonText) as StoryData;

      // Validate the story structure
      this.validateStoryData(storyData);

      const generationTime = Date.now() - startTime;
      console.log(`Story generated in ${generationTime}ms`);

      return storyData;
    } catch (error) {
      console.error('Error generating story:', error);
      if (error instanceof SyntaxError) {
        throw new Error('Failed to parse AI response as JSON');
      }
      throw error;
    }
  }

  private static validateStoryData(data: any): void {
    if (!data.title || typeof data.title !== 'string') {
      throw new Error('Invalid story: missing or invalid title');
    }
    if (!data.context || typeof data.context !== 'string') {
      throw new Error('Invalid story: missing or invalid context');
    }
    if (!data.character || typeof data.character !== 'object') {
      throw new Error('Invalid story: missing or invalid character');
    }
    if (!Array.isArray(data.scenes) || data.scenes.length === 0) {
      throw new Error('Invalid story: missing or invalid scenes');
    }

    // Validate each scene
    data.scenes.forEach((scene: any, index: number) => {
      if (!scene.id || !scene.dialogue || !scene.narration) {
        throw new Error(`Invalid scene ${index}: missing required fields`);
      }
      if (!Array.isArray(scene.choices) || scene.choices.length === 0) {
        throw new Error(`Invalid scene ${index}: missing or invalid choices`);
      }

      // Validate each choice
      scene.choices.forEach((choice: any, choiceIndex: number) => {
        if (!choice.id || !choice.text || typeof choice.isCorrect !== 'boolean') {
          throw new Error(`Invalid choice ${choiceIndex} in scene ${index}`);
        }
      });
    });
  }

  static async testConnection(): Promise<boolean> {
    try {
      const message = await client.messages.create({
        model: 'claude-sonnet-4-20250514',
        max_tokens: 100,
        messages: [
          {
            role: 'user',
            content: 'Hello',
          },
        ],
      });
      return message.content.length > 0;
    } catch (error) {
      console.error('Claude API connection test failed:', error);
      return false;
    }
  }
}

File: backend/app/services/claude_service.py
"""
Claude API integration service
"""
from anthropic import Anthropic
from typing import Dict, Any, List
import json
from ..config import get_settings


class ClaudeService:
    """
    Service for interacting with Claude AI API
    """

    def __init__(self):
        settings = get_settings()
        self.client = Anthropic(api_key=settings.anthropic_api_key)
        self.model = settings.claude_model

    async def analyze_learning_pattern(
        self,
        student_data: Dict[str, Any],
        activity_history: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """
        Analyze learning patterns and generate metacognitive insights

        Args:
            student_data: Student information
            activity_history: List of recent learning activities

        Returns:
            Dict containing analysis results and insights
        """
        prompt = self._build_analysis_prompt(student_data, activity_history)

        try:
            message = self.client.messages.create(
                model=self.model,
                max_tokens=2048,
                temperature=0.7,
                system=self._get_system_prompt(),
                messages=[
                    {
                        "role": "user",
                        "content": prompt
                    }
                ]
            )

            # Parse the response
            response_text = message.content[0].text
            return self._parse_analysis_response(response_text)

        except Exception as e:
            print(f"Error calling Claude API: {e}")
            return {
                "error": str(e),
                "insights": []
            }

    def _get_system_prompt(self) -> str:
        """Get the system prompt for Claude"""
        return """당신은 교육 심리학과 메타인지 전문가입니다.
학생의 학습 활동 데이터를 분석하여 메타인지 성장 포인트를 발견하고,
학생과 교사에게 유용한 인사이트를 제공하는 역할을 합니다.

메타인지 성장 영역:
1. 자기 조절 능력 (Self-Regulation): 힌트 의존도, 학습 계획, 집중도
2. 학습 효율성 (Learning Efficiency): 문제 해결 속도, 정확도 향상
3. 실수에서의 학습 (Learning from Errors): 오답 분석 및 개선
4. 자기 인식 (Self-Awareness): 자기평가 정확도, 난이도 인식
5. 학습 전략 발전 (Strategy Development): 접근 방식의 변화와 발전

분석 시 다음을 포함하세요:
- 구체적인 수치와 비교 데이터
- 긍정적이고 격려하는 톤
- 실행 가능한 조언
- 학생의 노력과 성장 강조

응답은 반드시 JSON 형식으로 작성하세요."""

    def _build_analysis_prompt(
        self,
        student_data: Dict[str, Any],
        activity_history: List[Dict[str, Any]]
    ) -> str:
        """Build the analysis prompt for Claude"""

        prompt = f"""다음 학생의 학습 데이터를 분석하여 오늘의 메타인지 성장 포인트를 추출해주세요.

## 학생 정보
- 이름: {student_data.get('name', 'Unknown')}
- 학년: {student_data.get('grade_level', 'Unknown')}

## 최근 학습 활동 데이터
{json.dumps(activity_history, ensure_ascii=False, indent=2)}

## 요청사항
위 데이터를 분석하여 다음 형식의 JSON으로 응답해주세요:

{{
  "insights": [
    {{
      "dimension": "자기_조절_능력|학습_효율성|실수에서의_학습|자기_인식|학습_전략_발전",
      "title": "성장 포인트 제목 (간결하게)",
      "description": "구체적인 설명 (수치 포함)",
      "recommendation": "실행 가능한 조언",
      "improvement_percentage": 개선 퍼센트 (숫자만),
      "confidence_score": 0.0-1.0,
      "evidence_data": {{
        "metric": "관련 지표",
        "previous_value": 이전 값,
        "current_value": 현재 값,
        "comparison_period": "비교 기간"
      }}
    }}
  ],
  "summary": "전체 학습 요약",
  "key_achievements": ["주요 성취 1", "주요 성취 2"],
  "recommendations": ["추천 사항 1", "추천 사항 2"]
}}

최소 3개, 최대 5개의 인사이트를 생성해주세요. 학생의 성장에 초점을 맞추고 긍정적인 톤을 유지해주세요."""

        return prompt

    def _parse_analysis_response(self, response_text: str) -> Dict[str, Any]:
        """
        Parse Claude's response into structured data

        Args:
            response_text: Raw text response from Claude

        Returns:
            Parsed dictionary
        """
        try:
            # Try to extract JSON from the response
            # Claude might wrap JSON in markdown code blocks
            if "```json" in response_text:
                json_start = response_text.find("```json") + 7
                json_end = response_text.find("```", json_start)
                json_text = response_text[json_start:json_end].strip()
            elif "```" in response_text:
                json_start = response_text.find("```") + 3
                json_end = response_text.find("```", json_start)
                json_text = response_text[json_start:json_end].strip()
            else:
                json_text = response_text.strip()

            parsed = json.loads(json_text)
            return parsed

        except json.JSONDecodeError as e:
            print(f"Error parsing Claude response: {e}")
            print(f"Response text: {response_text}")
            return {
                "error": "Failed to parse response",
                "raw_response": response_text,
                "insights": []
            }
